walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,163 @@
+   * @return If one or more time slots exists so that both mandatory and optional attendees can attend,
+   *         returns those time slots. Otherwise, returns the time slots that fit just the mandatory attendees.
+   *         Returns empty if duration is greater than a day.
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    long duration = request.getDuration();
+    if (!isValidParams(events, duration)) {
+      return Collections.unmodifiableList(new ArrayList<>());
+    }
+
+    Set<String> attendees = new HashSet<>(request.getAttendees());
+    Set<String> attendeesWithOptional = new HashSet<>();
+    attendeesWithOptional.addAll(request.getAttendees());
+    attendeesWithOptional.addAll(request.getOptionalAttendees());
+
+    // No attendees at all.
+    if (attendeesWithOptional.isEmpty()) {
+      return Collections.unmodifiableList(Arrays.asList(TimeRange.WHOLE_DAY));
+    }
+
+    List<TimeRange> occupiedTimeRange = new ArrayList<>();
+    List<TimeRange> occupiedTimeRangeWithOptional = new ArrayList<>();
+
+    for (Event event: events) {
+      getOccupiedTimeRange(attendees, occupiedTimeRange, event);
+
+      getOccupiedTimeRange(attendeesWithOptional, occupiedTimeRangeWithOptional, event);
+    }
+
+    List<TimeRange> availableMeetings = getAvailableTimeRange(occupiedTimeRangeWithOptional, duration);
+
+    // Has time slots that satisfy both mandatory and optional attendees.
+    if (!availableMeetings.isEmpty()) {
+      return Collections.unmodifiableList(availableMeetings);
+    }
+
+    // No mandatory attendees.
+    if (attendees.isEmpty()) {
+      return Collections.unmodifiableList(new ArrayList<>());
+    }
+
+    return Collections.unmodifiableList(getAvailableTimeRange(occupiedTimeRange, duration));
+  }
+
+  private boolean isValidParams(Collection<Event> events, long duration) {
+    // TODO: check for events that all events not crossing EOD.
+
+    if (duration > TimeRange.WHOLE_DAY.duration() || duration <= 0) {
+      return false;
+    }
+
+    return true;
+  }
+
+  private void getOccupiedTimeRange(Set<String> attendees, List<TimeRange> occupiedTimeRange, Event event) {
+    if (hasAttendeesAttending(event.getAttendees(), attendees)) {
+      occupiedTimeRange.add(event.getWhen());
+    }
+  }
+
+  private boolean hasAttendeesAttending(Set<String> eventAttendees, Set<String> requestAttendees) {
+    for (String attendee: requestAttendees) {
+      if (eventAttendees.contains(attendee)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  private List<TimeRange> getAvailableTimeRange(List<TimeRange> occupiedTimeRange, long duration) {
+    // Assumes 0 < duration <= WHOLE_DAY
+
+    List<TimeRange> availableMeetings = new ArrayList<>();
+
+    if (occupiedTimeRange.isEmpty()) {
+      availableMeetings.add(TimeRange.WHOLE_DAY);
+      return availableMeetings;
```

Comment:
Maybe directly return ImmutableList.of(TimeRange.WHOLE_DAY) ?

Reply:
change to `return Arrays.asList(TimeRange.WHOLE_DAY);` since it is only an internal private method